pandas/core/groupby/groupby.py

```diff
@@ -2250,6 +2251,22 @@ def post_processor(vals: np.ndarray, inference: Optional[Type]) -> np.ndarray:
             # reorder rows to keep things sorted
             indices = np.arange(len(result)).reshape([len(q), self.ngroups]).T.flatten()
             return result.take(indices)
+        else:
+            result = concat(results, axis=1, keys=q)
+
+            order = list(range(1, result.columns.nlevels)) + [0]
+            index_names = np.array(result.columns.names)
+            result.columns.names = np.arange(len(index_names))
+            result = result.reorder_levels(order, axis=1)
+            result.columns.names = index_names[order]
+            indices = (
+                np.arange(result.shape[1])
+                .reshape(
+                    [len(q), self.ngroups],
+                )
+                .T.flatten()
+            )
+            return result.take(indices, axis=1)
```

Comment:
could a lot of this be shared with the axis=0 version by replacing result.columns with `result.axes[1]` and above replacing `result.index` with `result.axes[0]`? (which in both cases would then become `result.axes[axis]`)

Reply:
sounds good, I will change this.